File: packages/soulseek_protocol/lib/src/services/chat_service.dart
import 'dart:async';

import '../connection/server_connection.dart';
import '../messages/buffer.dart';
import '../messages/codes.dart';
import '../messages/message.dart';

class ChatService {
  final ServerTransport _server;
  final _privateMessageController = StreamController<PrivateMessage>.broadcast();

  late final StreamSubscription _messageSub;

  ChatService({required ServerTransport server}) : _server = server;

  Stream<PrivateMessage> get privateMessages => _privateMessageController.stream;

  void init() {
    _messageSub = _server.messages.listen(_onMessage);
  }

  void sendPrivateMessage(String username, String message) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.privateMessage,
      PrivateMessage(
        username: username,
        message: message,
        timestamp: DateTime.now().millisecondsSinceEpoch ~/ 1000,
      ).serialize().toBytes(),
    ));
  }

  void _onMessage(SoulseekMessage message) {
    switch (message.code) {
      case ServerCode.privateMessage:
        _handlePrivateMessage(message);
        break;
    }
  }

  void _handlePrivateMessage(SoulseekMessage message) {
    try {
      final buffer = ReadBuffer(message.payload);
      final pm = PrivateMessage.parse(buffer);
      _privateMessageController.add(pm);
    } catch (_) {
      // Skip malformed messages
    }
  }

  void dispose() {
    _messageSub.cancel();
    _privateMessageController.close();
  }
}

File: Soulseek.Protocol/Services/ChatService.cs
using Soulseek.Protocol.Connection;
using Soulseek.Protocol.Messages;

namespace Soulseek.Protocol.Services;

public class ChatService
{
    private readonly IServerTransport _server;
    private readonly Subject<PrivateMessage> _privateMessageSubject = new();
    private IDisposable? _messageSub;

    public ChatService(IServerTransport server)
    {
        _server = server;
    }

    public IObservable<PrivateMessage> PrivateMessages => _privateMessageSubject;

    public void Init()
    {
        _messageSub = _server.Messages.Subscribe(OnMessage);
    }

    public void SendPrivateMessage(string username, string message)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.PrivateMessage,
            new PrivateMessage(
                username,
                message,
                (int)DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            ).Serialize().ToBytes()
        ));
    }

    private void OnMessage(SoulseekMessage message)
    {
        if (message.Code != ServerCode.PrivateMessage) return;

        try
        {
            var buffer = new ReadBuffer(message.Payload);
            var pm = PrivateMessage.Parse(buffer);
            _privateMessageSubject.OnNext(pm);
        }
        catch
        {
            // Skip malformed messages
        }
    }

    public void Dispose()
    {
        _messageSub?.Dispose();
    }
}